File: java/pen/Pen.java
public class Pen implements Writable {
    protected int amount;
    protected int capacity;

    public Pen(int capacity) {
        this.capacity = capacity;
        this.amount = capacity;
    }

    public Pen() {
        this(4096);
    }

    public int getAmount() {
        return this.amount;
    }
    public int getCapacity() {
        return this.capacity;
    }

    public void refill() {
        this.amount = this.capacity;
    }

    public void write(Paper paper, String message) throws
        OutOfInkException, OutOfSpaceException, ClosedPenException
    {
        if ( this.amount == 0 ) {
            throw new OutOfInkException();
        }
        if ( message.length() > this.amount ) {
            paper.addContent(message.substring(0, this.amount));
            this.amount = 0;
            return;
        }
        paper.addContent(message);
        this.amount -= message.length();
    }

    public String toString() {
        StringBuilder out = new StringBuilder("Pen(");

        out.append(this.amount);
        out.append('/');
        out.append(this.capacity);
        out.append(')');

        return out.toString();
    }
}

File: pen/Pen.cpp
#include "Pen.hpp"
#include "exception.hpp"

Pen::Pen(int capacity, const char* title)
    : amount(capacity), capacity(capacity), title(title) {}

Pen::~Pen() {}

int Pen::getAmount() const {
    return this->amount;
}

int Pen::getCapacity() const {
    return this->capacity;
}

const char* Pen::getTitle() const {
    return this->title;
}

void Pen::refill() {
    this->amount = this->capacity;
}

void Pen::write(Paper& paper, const std::string& message) {
    if ( this->amount == 0 ) {
        throw OutOfInkException();
    }
    if ( this->amount < message.length() ) {
        paper.addContent(message.substr(0, this->amount));
        this->amount = 0;
        return;
    }
    paper.addContent(message);
    this->amount -= message.length();
}

std::ostream& operator<<(std::ostream& out, const Pen& pen) {
    out << pen.getTitle() << "(" << pen.getAmount() << "/";
    out << pen.getCapacity() << ")";

    return out;
}
